src/Exchange.js

```diff
@@ -36,37 +43,208 @@ export default function Exchange(props) {
-        BigInt(inputLiquidityToken.totalSupply),
-        BigInt(inputLiquidityToken.price)
-      ),
-      BASE_FACTOR
+
+  const [exchange] = usePostTransaction({
+    contract: "Exchange",
+    functionName: "exchange",
+  });
+  const clearForm = () => {
+    setInputAmount("");
+    setError("");
+    setMinimumOutputAmount("0");
+  };
+
+  const calculateAvailableQuantity = (
+    inputLiquidityToken,
+    outputLiquidityToken,
+    outputTokenName
+  ) => {
+    const quantity =
+      outputTokenName === "USD"
+        ? (inputLiquidityToken.totalPoolSupply / inputLiquidityToken.price) *
+          BASE_FACTOR
+        : outputLiquidityToken.totalPoolSupply / BASE_FACTOR;
+
+    setAvailableQuantity(isNaN(quantity) ? new BigInt(0) : quantity);
+  };
+
+  const setFeeInInputToken = (
+    feeInBaseToken,
+    priceInBaseToken,
+    inputTokenName
+  ) => {
+    setFee(
+      inputTokenName === "USD"
+        ? feeInBaseToken
+        : feeInBaseToken / priceInBaseToken
     );
-    const invarient = multiply(elcPool, usdPool);
-    const newElcPool = subtract(add(elcPool, inputAmount), fee);
-    const newUsdPool = divide(invarient, newElcPool);
-    const outputAmount = subtract(usdPool, newUsdPool);
-    console.log(outputToken.name);
+  };
+
+  const applyFee = (amount) => {
+    return subtract(
+      amount,
+      divide(multiply(amount, LIQUIDITY_FEE), BASE_FACTOR)
+    );
+  };
+
+  const getBaseTokenReserves = (totalSupply, price) => {
+    return divide(multiply(totalSupply, price), BASE_FACTOR);
+  };
+
+  React.useEffect(() => {
+    const _inputLiquidityToken = find(liquidityTokens, ["id", inputToken.id]);
+    setInputLiquidityToken(_inputLiquidityToken);
+
+    const _outputLiquidityToken = find(liquidityTokens, ["id", outputToken.id]);
+    setOutputLiquidityToken(_outputLiquidityToken);
+
+    const userToken = find(userTokens, ["id", inputToken.id]);
+    setUserTokenBalance(userToken.balance);
+
+    calculateAvailableQuantity(
+      inputLiquidityToken,
+      outputLiquidityToken,
+      outputToken.name
+    );
+  }, [
+    inputToken,
+    outputToken,
+    userTokens,
+    inputLiquidityToken,
+    outputLiquidityToken,
+    liquidityTokens,
+  ]);
+
+  const exchangeRate = useMemo(() => {
+    if (inputToken.ticker === outputToken.ticker) {
+      return 1;
+    }
+
+    let outputTokenPrice =
+      outputToken.ticker === "USD"
+        ? BASE_FACTOR
+        : outputLiquidityToken.price || 0;
+    let inputTokenPrice =
+      inputToken.ticker === "USD"
+        ? BASE_FACTOR
+        : inputLiquidityToken.price || 0;
+    if (outputTokenPrice !== 0) {
+      return inputTokenPrice / outputTokenPrice;
+    } else {
+      return null;
+    }
+  }, [inputToken, outputToken, inputLiquidityToken, outputLiquidityToken]);
+
+  const outputAmount = useMemo(() => {
+    const calculateInputAmountInBaseToken = (amount, totalSupply, price) => {
+      if (Number(totalSupply) === 0) return new BigInt(0);
```

Comment:
Any reason to use `Number(totalSupply)` instead of  `BigInt(0)`. For the most part we should use `BigInt`s to prevent overflow. I think we're going to be safe for the most part but probably worth it just in case...

Reply:
Yeah, unfrotunately `if (totalSupply === new BigInt(0))` evaluates to false when `totalSupply` _is_ `0`.

Reply:
ah! Importing `equals` from `JSBI` and using that should work

Reply:
Haha 🤦